File: Stack/minimum_stack.java
import java.util.Stack;

public class minimum_stack {
    static class MinStack {
        private Stack<Integer> stack;
        private Stack<Integer> minStack;
        
        public MinStack() {
            // we need two stacks one normal, and one for minimum to get the minimum in O(1) time
            stack = new Stack<>();
            minStack = new Stack<>();
        }

        public void push(int val) {
            stack.push(val); // push normally to this stack
            val = Math.min(val, minStack.isEmpty() ? val : minStack.peek()); // get the minimum 
            minStack.push(val); // push that min val to the minStack
        }

        public void pop() { // just pop from both
            stack.pop();
            minStack.pop();
        }

        public int top() { // use peek function as it returns the top of the stack without removing it
            return stack.peek();
        }

        public int getMin() { // use peek function to return the top as in the minStack top value will always be the minimum
            return minStack.peek();
        }

        public void printStack() {
            System.out.println(stack);
            System.out.println(minStack);
        }
    }
    public static void main(String[] args) {
        MinStack minStack = new MinStack();

        minStack.push(1);
        minStack.printStack();
        minStack.push(2);
        minStack.printStack();
        minStack.push(0);
        minStack.printStack();
        System.out.println(minStack.getMin());
        minStack.pop();
        minStack.printStack();
        System.out.println(minStack.top());
        System.out.println(minStack.getMin());

        // time complexity will be O(1) for all functions but space complexity can be O(n) depending on size of stack
    } 
}

File: Stack/MinimumStack.py
# Time Complexity: O(1) for all functions
# Space Complexity: O(n) for both min stacks
'''
The time complexity for all functions is O(1) as it should be, but the space complexity is O(n) because the size 
of the stacks will increase as more integers are pushed in
'''


class MinStack:

    def __init__(self):
        self.minStack = []
        self.minStack2 = []
        
    def push(self, val: int) -> None:
        
        self.minStack.append(val)

        self.minStack2.append(min(val, self.minStack2[-1] if self.minStack2 else val))

    def pop(self) -> None:
        self.minStack.pop()
        self.minStack2.pop()
        

    def top(self) -> int:
        return self.minStack[-1]
        

    def getMin(self) -> int:
        return self.minStack2[-1]
    

minStack = MinStack()
minStack.push(1)
minStack.push(2)
minStack.push(0)
minStack.getMin()
minStack.pop()
minStack.top()    
minStack.getMin()